src/ccl_correlation.c

```diff
@@ -506,3 +509,419 @@ void ccl_correlation_3d(ccl_cosmology *cosmo, double a,
+    if (xi_arr2 == NULL) {
+      free(xi_arr0);
+      *status = CCL_ERROR_MEMORY;
+      strcpy(cosmo->status_message,
+             "ccl_correlation.c: ccl_correlation_3dRsd ran out of memory\n");
+      return;
+    }
+    xi_arr4 = malloc(sizeof(double) * n_s);
+    if (xi_arr4 == NULL) {
+      free(xi_arr2);
+      free(xi_arr4);
+      *status = CCL_ERROR_MEMORY;
+      strcpy(cosmo->status_message,
+             "ccl_correlation.c: ccl_correlation_3dRsd ran out of memory\n");
+      return;
+    }
+
+    ccl_correlation_multipole(cosmo, a, beta, 0, n_s, s, xi_arr0, status);
+    ccl_correlation_multipole(cosmo, a, beta, 2, n_s, s, xi_arr2, status);
+    ccl_correlation_multipole(cosmo, a, beta, 4, n_s, s, xi_arr4, status);
+    for (i = 0; i < n_s; i++)
+      xi[i] = xi_arr0[i] + xi_arr2[i] * gsl_sf_legendre_Pl(2, mu) +
+              xi_arr4[i] * gsl_sf_legendre_Pl(4, mu);
+    free(xi_arr0);
+    free(xi_arr2);
+    free(xi_arr4);
+
+  } else {
+    if (xir_spline[0] == NULL)
+      ccl_correlation_multipole_spline(cosmo, a, status);
+
+    for (i = 0; i < n_s; i++)
+      xi[i] = (1. + 2. / 3 * beta + 1. / 5 * beta * beta) *
+                  ccl_spline_eval(s[i], xir_spline[0]) -
+              (4. / 3 * beta + 4. / 7 * beta * beta) *
+                  ccl_spline_eval(s[i], xir_spline[1]) *
+                  gsl_sf_legendre_Pl(2, mu) +
+              8. / 35 * beta * beta * ccl_spline_eval(s[i], xir_spline[2]) *
+                  gsl_sf_legendre_Pl(4, mu);
+  }
+
+  ccl_check_status(cosmo, status);
+
+  return;
+}
+
+void ccl_correlation_3dRsd_avgmu(ccl_cosmology *cosmo, double a, int n_s, double *s,
+                                 double beta, double *xi,
+                                 int *status) {
+  int i;
+  double *xi_arr0, *xi_arr2, *xi_arr4;
+
+    xi_arr0 = malloc(sizeof(double) * n_s);
+    if (xi_arr0 == NULL) {
+      *status = CCL_ERROR_MEMORY;
+      strcpy(cosmo->status_message,
+             "ccl_correlation.c: ccl_correlation_3dRsd_avgmu ran out of memory\n");
+      return;
+    }
+    xi_arr2 = malloc(sizeof(double) * n_s);
+    if (xi_arr2 == NULL) {
+      free(xi_arr0);
+      *status = CCL_ERROR_MEMORY;
+      strcpy(cosmo->status_message,
+             "ccl_correlation.c: ccl_correlation_3dRsd_avgmu ran out of memory\n");
+      return;
+    }
+    xi_arr4 = malloc(sizeof(double) * n_s);
+    if (xi_arr4 == NULL) {
+      free(xi_arr2);
+      free(xi_arr4);
+      *status = CCL_ERROR_MEMORY;
+      strcpy(cosmo->status_message,
+             "ccl_correlation.c: ccl_correlation_3dRsd_avgmu ran out of memory\n");
+      return;
+    }
+
+    ccl_correlation_multipole(cosmo, a, beta, 0, n_s, s, xi_arr0, status);
+    ccl_correlation_multipole(cosmo, a, beta, 2, n_s, s, xi_arr2, status);
+    ccl_correlation_multipole(cosmo, a, beta, 4, n_s, s, xi_arr4, status);
+    for (i = 0; i < n_s; i++)
+      xi[i] = xi_arr0[i] + xi_arr2[i] * 1./4 +
+              xi_arr4[i] * 9./64;
```

Comment:
Huhm, are these factors correct? Perhaps yes, but I don't recognize the numbers.

Reply:
Yes these are correct, but different from the coefficients in the expressions for xi(s,mu) that you are probably thinking of. This code is calculating xi(s) = average of xi(s,mu) at constant s. So here we calculate the multipoles  and multiply by the average of L_l(mu). So these coefficients are essentially the integrals of the Legendre polynomials.

Reply:
@slosar. Actually, we double checked this and found that this is the average over theta but we want the average over mu, so the second and third coefficients should be zero.  In this case (since we are using the linear approximation) xi(s) is just related to xi(r)  by xi(s) = (1 + 2/3 *beta + beta^2/5) xi(r). We will fix this.

Reply:
@slosar Actually, we double-checked and these are the coefficients for averaging over theta but we want to average over mu. In this case the second and third coefficients are zero (since we are using the linear Kaiser approximation) and xi(s) is related to xi(r) by xi(s) = (1 + 2/3 beta + (beta^2)/5) xi(r).   We have fixed this now.

The agreement with the benchmark is now much better, see the plots attached.

The maximum difference in s^2 xi(s) is 0.058, which is 0.11% of the BAO peak height.

Reply:
Yes, I recognize _these_ numbers!